lib/mzgl/gl/api/OpenGLAPI.h

```diff
@@ -0,0 +1,337 @@
+#pragma once
+
+#include "GraphicsAPI.h"
+#include "mzOpenGL.h"
+
+class OpenGLAPI : public GraphicsAPI {
+public:
+	OpenGLAPI(Graphics &g)
+		: GraphicsAPI(g) {}
+
+	void init() override {
+		glGetIntegerv(GL_FRAMEBUFFER_BINDING, &g.defaultFBO);
+
+		const unsigned char *openglVersion = glGetString(GL_VERSION);
+		if (openglVersion != nullptr) {
+			Log::d() << "OpenGL Version: " << openglVersion;
+#ifdef MZGL_GL2
+			Log::d() << "MZGL Compiled with MZGL_GL2==TRUE";
+#else
+			Log::d() << "MZGL Compiled with MZGL_GL2==FALSE";
+#endif
+		} else {
+			Log::d() << "OpenGL Version: null";
+		}
+		loadDefaultShaders();
+
+#ifndef MZGL_GL2
+		if (immediateVertexArray != 0) {
+			Log::e()
+				<< "Immediate vertex array recreated - is this bad? if on android, probs should clean this up";
+		}
+		glGenVertexArrays(1, &immediateVertexArray);
+		glBindVertexArray(immediateVertexArray);
+
+#endif
+		if (immediateVertexBuffer != 0) {
+			Log::e() << "Immediate vertex buffer recreated - is this bad?";
+		}
+
+		glGenBuffers(1, &immediateVertexBuffer);
+		glGenBuffers(1, &immediateColorBuffer);
+		glGenBuffers(1, &immediateIndexBuffer);
+	}
+
+	void loadDefaultShaders() {
+		// zero out all shaders before loading any, for android mostly,
+		// as the context may have reset itself
+		g.nothingShader		 = nullptr;
+		g.colorShader		 = nullptr;
+		g.colorTextureShader = nullptr;
+		g.fontShader		 = nullptr;
+		g.texShader			 = nullptr;
+
+		// clang-format off
+		g.nothingShader = Shader::create(g);
+
+		g.nothingShader->loadFromString(STRINGIFY(
+			uniform mat4 mvp;
+
+			in vec4 Position;
+			uniform lowp vec4 color;
+			out lowp vec4 colorV;
+
+			void main(void) {
+				colorV	  = color;
+				gl_Position = mvp * Position;
+			}
+
+		),
+		STRINGIFY(
+			in lowp vec4 colorV;
+			out vec4 fragColor;
+
+			void main(void) {
+				fragColor = colorV;
+			}
+		));
+
+		g.colorShader = Shader::create(g);
+		g.colorShader->loadFromString(STRINGIFY(
+
+			uniform mat4 mvp;
+			uniform lowp vec4 color;
+
+			in vec4 Position;
+			in lowp vec4 Color;
+
+			out lowp vec4 colorV;
+
+			void main(void) {
+				colorV	  = Color * color;
+				gl_Position = mvp * Position;
+			}
+
+		),
+
+		STRINGIFY(
+
+			in lowp vec4 colorV;
+			out vec4 fragColor;
+
+			void main(void) {
+				fragColor = colorV;
+			}
+
+		));
+
+		g.colorTextureShader = Shader::create(g);
+		g.colorTextureShader->loadFromString(STRINGIFY(
+
+			uniform mat4 mvp;
+
+			in vec4 Position;
+			in lowp vec2 TexCoord;
+			in lowp vec4 Color;
+
+			out lowp vec4 colorV;
+			out lowp vec2 texCoordV;
+			void main(void) {
+				colorV		= Color;
+				texCoordV	= TexCoord;
+				gl_Position = mvp * Position;
+			}
+
+		),
+STRINGIFY(
+
+			in lowp vec4 colorV;
+			in lowp vec2 texCoordV;
+			out vec4 fragColor;
+			uniform sampler2D myTextureSampler;
+
+			void main(void) {
+				fragColor = texture(myTextureSampler, texCoordV) * colorV;
+			}
+
+		));
```

Comment:
One way ive done this before is to have the strings in a separate file, and then just #include them in the code here. that way the shader code doesn't get obscured by the actual shader strings. At the moment the actual function and its behaviour is a bit hard to follow because of the the STRINGIFY parts

Reply:
good point - this will all have to change with the new shader api, but for now I will sling these in a separate file in any case